test/hotspot/jtreg/runtime/cds/appcds/dynamicArchive/TestAutoCreateSharedArchiveUpgrade.java

```diff
@@ -61,21 +72,33 @@ public class TestAutoCreateSharedArchiveUpgrade {
     private static String newJVM;
 
     public static void main(String[] args) throws Throwable {
-        setupJVMs();
-        doTest();
+        // Get OS and CPU type
+        String arch = props.getProperty("os.arch");
+        String os = getOsId();
+
+        // Earliest testable version is 19
+        int n = java.lang.Runtime.version().major() - 1;
+        for (int i = 19; i < n; i++) {
+            BOOT_JDK = fetchBootJDK(os, arch, i);
+            setupJVMs(os);
+            doTest();
+        }
     }
 
-    static void setupJVMs() throws Throwable {
+    static void setupJVMs(String os) throws Throwable {
         if (TEST_JDK == null) {
             throw new RuntimeException("-Dtest.jdk should point to the JDK being tested");
         }
 
         newJVM = TEST_JDK + FS + "bin" + FS + "java";
 
+        // Example path: bundles/linux-x64/jdk-19_linux-x64_bin.tar.gz/jdk-19/bin/java
         if (PREV_JDK != null) {
             oldJVM = PREV_JDK + FS + "bin" + FS + "java";
         } else if (BOOT_JDK != null) {
-            oldJVM = BOOT_JDK + FS + "bin" + FS + "java";
+            oldJVM = (os == "MacOSX") ?
+                BOOT_JDK + ".jdk" + FS + "Contents" + FS + "Home" + FS + "bin" + FS + "java" :
+                BOOT_JDK + FS + "bin" + FS + "java";
```

Comment:
As mentioned in a comment in [JDK-8296754](https://bugs.openjdk.org/browse/JDK-8296754), we should change the RuntimeException in the next block to a SkippedException.